app/controllers/application_controller/filter.rb

```diff
@@ -0,0 +1,152 @@
+            send(type, k, v)
+          else
+            errors.add("unsupported attribute type for: #{k}")
+          end
+        end
+      end
+
+      raise ApplicationController::Filter::Error.new(:error_document => errors) unless errors.blank?
+      query
+    end
+
+    private
+
+    attr_accessor :query
+
+    class Error < ArgumentError
+      attr_reader :error_document
+
+      def initialize(attrs)
+        @error_document = attrs[:error_document]
+      end
+    end
+
+    def with_attribute_for_key(key)
+      attribute = api_doc_definition.properties[key.to_s]
+      attribute ? yield(attribute) : errors.add("found unpermitted parameter: #{key}")
+    end
+
+    def determine_string_attribute_type(attribute)
+      return :timestamp if attribute["format"] == "date-time"
+      return :integer if attribute["pattern"] == /^\d+$/
+      :string
+    end
+
+    def errors
+      @errors ||= TopologicalInventory::Api::ErrorDocument.new
+    end
+
+    def string(k, val)
+      if val.kind_of?(ActionController::Parameters)
+        val.each do |comparator, value|
+          add_filter(comparator, STRING_COMPARISON_KEYWORDS, k, value)
+        end
+      else
+        add_filter("eq", STRING_COMPARISON_KEYWORDS, k, val)
+      end
+    end
+
+    def add_filter(requested_comparator, allowed_comparators, key, value)
+      type = with_attribute_for_key(key) { |attribute| determine_string_attribute_type(attribute) }
+
+      if requested_comparator.in?(["nil", "not_nil"])
+        send("comparator_#{requested_comparator}", key, value)
+      elsif requested_comparator.in?(allowed_comparators)
+        value = parse_datetime(value) if type == :datetime
+        return if value.nil?
+        send("comparator_#{requested_comparator}", key, value)
+      else
+        errors.add("unsupported #{type} comparator: #{requested_comparator}")
+      end
+    end
+
+    def timestamp(k, val)
+      if val.kind_of?(ActionController::Parameters)
+        val.each do |comparator, value|
+          add_filter(comparator, INTEGER_COMPARISON_KEYWORDS, k, value)
+        end
+      else
+        add_filter("eq", INTEGER_COMPARISON_KEYWORDS, k, val)
+      end
+    end
+
+    def parse_datetime(value)
+      return value.collect { |i| parse_datetime(i, ) } if value.kind_of?(Array)
+
+      DateTime.parse(value)
+    rescue ArgumentError
+      errors.add("invalid timestamp: #{value}")
+      return nil
+    end
+
+    def integer(k, val)
+      if val.kind_of?(ActionController::Parameters)
+        val.each do |comparator, value|
+          add_filter(comparator, INTEGER_COMPARISON_KEYWORDS, k, value)
+        end
+      else
+        add_filter("eq", INTEGER_COMPARISON_KEYWORDS, k, val)
+      end
+    end
+
+    def comparator_contains(key, value)
+      return value.each { |v| comparator_contains(key, v) } if value.kind_of?(Array)
+      self.query = query.where("#{key} LIKE ?", "%#{value}%")
```

Comment:
we should probably quote the key and elsewhere...either that or use `?` (though I'm not sure you *can* use `?` in the key part.  I'm concerned about possible injection with this via query params.

Reply:
I think the `key` is safe here since `with_attribute_for_key` will error out if the exact key is not found in the swagger doc.

Reply:
Oh, I see, that should be fixed now.